doc: add general ZMS info (#18174)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/nrf54l/non-volatile_storage.rst

﻿.. _config_settings_zms:

Non-volatile storage
####################

.. contents::
   :local:
   :depth: 2

For nRF54L Series, you should use the :ref:`Zephyr Memory Storage (ZMS) <zephyr:zms_api>`.

> ZMS utilizes a flexible data management system that reduces write and erase cycles, extending the lifespan of flash memory.

Reviewer: ```suggestion
ZMS utilizes a flexible data management system that reduces write and erase cycles, extending the lifespan of non-volatile memory.
```

nRF54L has RRAM rather than flash.

Author: fixed

.. note::

  If you are using the :ref:`Non-Volatile Storage (NVS) <zephyr:nvs_api>`, ensure to switch to the ZMS solution.

Enabling the ZMS module
***********************

You can enable the ZMS module for your application in two ways - directly through ZMS or through the Settings subsystem.
Additionally, you can optimize the application performance using caching mechanism:

.. tabs::

   .. group-tab:: Using storage directly through ZMS

      1. Enable the :kconfig:option:`CONFIG_ZMS` Kconfig option in your project's configuration file.
      #. Optionally, enable the lookup cache to increase the application performance:

        a. Enable the :kconfig:option:`CONFIG_ZMS_LOOKUP_CACHE` Kconfig option.
        #. Set the lookup cache size in :kconfig:option:`CONFIG_ZMS_LOOKUP_CACHE_SIZE` to ``512``.

   .. group-tab:: Using storage through Settings subsystem

>       1. Enable the :kconfig:option:`CONFIG_ZMS` configuration.

Reviewer 2: ```suggestion
      1. Enable the :kconfig:option:`CONFIG_ZMS` Kconfig option in your project's configuration file.
```

      #. Enable ZMS within the Settings subsystem with the :kconfig:option:`CONFIG_SETTINGS_ZMS` Kconfig option.
      #. Optionally, enable the lookup cache to increase the application performance:

        a. Enable the :kconfig:option:`CONFIG_ZMS_LOOKUP_CACHE` Kconfig option.
        #. Set the lookup cache size in :kconfig:option:`CONFIG_ZMS_LOOKUP_CACHE_SIZE` to ``512``.
        #. Ensure the lookup cache is configured to support the Setting subsystem by enabling the :kconfig:option:`CONFIG_ZMS_LOOKUP_CACHE_FOR_SETTINGS` Kconfig option.

Optimizing ZMS in your application
**********************************

> When integrating ZMS through the Settings subsystem in your application, you might need to optimize its performance, especially when handling a significant number of entries.

Reviewer: How many entries is significant?

The following sections include the best practices to enhance ZMS performance.

Sector size and count
=====================

> To ensure optimal performance from ZMS, you should properly dimension both the total storage partition size and the sector size:

Reviewer 2: ```suggestion
To ensure optimal performance from ZMS, consider the following:
```
Not 100% happy with this wording, but it matches the bullets better.

> * Storage partition sizing - The total storage partition size and the sector size should be carefully dimensioned to optimize ZMS performance.

Reviewer 2: For all bullets, I'd add an empty line after the first line, to separate the additional information to its own paragraph in the bullet. But that's a personal preference, not a necessary change.
